plugins/inputs/aliyuncms/aliyuncms_test.go

```diff
@@ -0,0 +1,485 @@
+		metricName          string
+		expectedErrorString string
+	}{
+		{
+			name:                "404 error from CMS endpoint",
+			metricName:          "ErrorCode",
+			expectedErrorString: "failed to query metricName list: ErrorCode",
+		},
+		{
+			name:                "Datapoint with corrupted JSON",
+			metricName:          "ErrorDatapoint",
+			expectedErrorString: `failed to decode response datapoints: invalid character '}' looking for beginning of object key string`,
+		},
+		{
+			name:                "General CMS response error",
+			metricName:          "ErrorResp",
+			expectedErrorString: "failed to query metricName list: error response",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var acc telegraf.Accumulator
+			require.EqualError(t, plugin.gatherMetric(acc, tt.metricName, metric), tt.expectedErrorString)
+
+		})
+	}
+}
+
+func TestGather(t *testing.T) {
+
+	metric := &Metric{
+		MetricNames: []string{},
+		Dimensions:  `{"instanceId": "i-abcdefgh123456"}`,
+	}
+	plugin := &AliyunCMS{
+		AccessKeyID:      "my_access_key_id",
+		AccessKeySecret:  "my_access_key_secret",
+		Project:          "acs_slb_dashboard",
+		Metrics:          []*Metric{metric},
+		RateLimit:        200,
+		measurement:      formatMeasurement("acs_slb_dashboard"),
+		DiscoveryRegions: []string{"cn-shanghai"},
+		client:           new(mockGatherAliyunCMSClient),
+	}
+
+	//test table:
+	tests := []struct {
+		name          string
+		hasMeasurment bool
+		metricNames   []string
+		expected      []telegraf.Metric
+	}{
+		{
+			name:        "Empty data point",
+			metricNames: []string{"EmptyDatapoint"},
+			expected: []telegraf.Metric{
+				testutil.MustMetric(
+					"aliyuncms_acs_slb_dashboard",
+					nil,
+					nil,
+					time.Time{}),
+			},
+		},
+		{
+			name:          "Data point with fields & tags",
+			hasMeasurment: true,
+			metricNames:   []string{"InstanceActiveConnection"},
+			expected: []telegraf.Metric{
+				testutil.MustMetric(
+					"aliyuncms_acs_slb_dashboard",
+					map[string]string{
+						"instanceId": "i-abcdefgh123456",
+						"userId":     "1234567898765432",
+					},
+					map[string]interface{}{
+						"instance_active_connection_minimum": float64(100),
+						"instance_active_connection_maximum": float64(200),
+						"instance_active_connection_average": float64(150),
+						"instance_active_connection_value":   float64(300),
+					},
+					time.Unix(1490152860000, 0)),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var acc testutil.Accumulator
+			plugin.Metrics[0].MetricNames = tt.metricNames
+			require.Empty(t, acc.GatherError(plugin.Gather))
+			require.Equal(t, acc.HasMeasurement("aliyuncms_acs_slb_dashboard"), tt.hasMeasurment)
+			if tt.hasMeasurment {
+				acc.AssertContainsTaggedFields(t, "aliyuncms_acs_slb_dashboard", tt.expected[0].Fields(), tt.expected[0].Tags())
+			}
+		})
+	}
+}
+
+//func TestParseDiscoveryResponse(t *testing.T) {
```

Comment:
should probably fix or remove the commented-out test

Reply:
Agree, removed